### ISSUE TITLE
"pickle.Unpickler.persistent_load" and "pickle.Pickler.persistent_id" became read-only in python 3.13

### ISSUE THREAD

**Issue by jcea** [python/cpython#125631]
# Bug report

### Bug description:

In Python releases previous to 3.13, the following code used to work (this idiom is used in Durus and ZODB persistence systems):

```
Python 3.12.7 (main, Oct  3 2024, 03:21:52) [GCC 10.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pickle import Unpickler
>>> a=Unpickler(open('/dev/zero'))  
>>> a.persistent_load=lambda x: x
>>> 
```

Fine so far.

In Python 3.13, this doesn't work anymore:

```
Python 3.13.0 (main, Oct  9 2024, 14:54:06) [GCC 10.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pickle import Unpickler
>>> a=Unpickler(open('/dev/zero'))
>>> a.persistent_load=lambda x: x
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    a.persistent_load=lambda x: x
    ^^^^^^^^^^^^^^^^^
AttributeError: '_pickle.Unpickler' object attribute 'persistent_load' is read-only
```

I don't know if this is an intended change or a regression.

PS: The subclass approach works:

```
Python 3.13.0 (main, Oct  9 2024, 14:54:06) [GCC 10.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pickle import Unpickler
>>> def x(Unpickler):
...     def persistent_load(self, x):
...         return x
...         
>>> b=x(open("/dev/zero"))
>>> 
```

### CPython versions tested on:

3.13

### Operating systems tested on:

Linux

### Linked PRs
* gh-125724
* gh-125752
* gh-126528

**Comment by jcea** [comment]
The same issue happens with ```pickle.Pickler.persistent_id```:

```
Python 3.12.7 (main, Oct  3 2024, 03:21:52) [GCC 10.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pickle import Pickler
>>> a=Pickler(open('/dev/zero'))
>>> a.persistent_id=lambda x: x
>>> 
```

Good so far.

But in Python 3.13.0:

```
Python 3.13.0 (main, Oct  9 2024, 14:54:06) [GCC 10.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pickle import Pickler
>>> a=Pickler(open('/dev/zero'))
>>> a.persistent_id=lambda x: x
Traceback (most recent call last):
  File "<python-input-6>", line 1, in <module>
    a.persistent_id=lambda x: x
    ^^^^^^^^^^^^^^^
AttributeError: '_pickle.Pickler' object attribute 'persistent_id' is read-only
>>> 
```

**Comment by jcea** [comment]
It seems related to bug #89850 and PR  #113579.

**Comment by jcea** [comment]
@serhiy-storchaka, Could you possibly verify this? Thanks.

**Comment by serhiy-storchaka** [comment]
Yes, this is a consequence of the #89850 change. I missed such use case because there were no tests for it. You still can override attributes of instances of subclasses.

The proposed fix adds managed dicts to these classes (I would add simple instance dicts, but it is easier to add managed dicts). This may open a can of worms, because you can now set arbitrary attributes of instances of these classes. There will be no way to close it. But I think this is the only way to make this working while keeping `super()` working too.

**Comment by jcea** [comment]
What about restoring the property support (get/set) for those two attributes?

It could even support "delete" to reactivate current configuration, although I don't think that anybody overwriting "persistent_id" or "persistent_load" would be interested in that.

**Comment by serhiy-storchaka** [comment]
This would be very complex solution and it would not work with `super()`.

But we can make attributes (and only these attributes) writable by implementing `tp_getattro` and `tp_setattro`. This is slightly more complex solution, but not extremely complex. See #125752.

**Comment by paulankita137** [comment]
Does this problem have a fix ? I am facing the same issues

**Comment by jcea** [comment]
It was solved in Python 3.13.1, already a couple of weeks old.
